enhancements/microshift/microshift-telemetry.md

```diff
@@ -0,0 +1,277 @@
+to MicroShift.
+
+## Motivation
+MicroShift is currently not sending any kind of metric from production
+deployments. Not having remote monitoring creates a blind spot on the
+deployment and usage characteristics of MicroShift instances in production
+environments.
+
+MicroShift runs as an application on top of R4E (Rhel4Edge), the usage of RHEL
+insights may yield some information about MicroShift, but there might be gaps.
+RHEL insights knows about which packages are installed, but not whether
+MicroShift is running or runtime metrics.
+
+In order to enhance the user experience and Red Hat's insights about production
+MicroShift deployments, this enhancement proposes enabling the use of Telemetry
+API to get data from connected MicroShift clusters.
+
+### User Stories
+As Red Hat, I want to enable MicroShift clusters to report back to me to get
+knowledge on usage patterns of live deployments.
+
+As a MicroShift admin, I want to have the option to opt-out of telemetry.
+
+As a MicroShift admin, I want to have the option to configure how often
+telemetry data is sent.
+
+
+### Goals
+* Enable MicroShift connected deployments to send current information about
+system capabilities and usage characteristics.
+
+* Get better understanding of deployment and usage patterns on MicroShift.
+
+### Non-Goals
+* Have an in-cluster metrics service.
+
+* Provide recommendations and/or analysis to customers using their data.
+
+## Proposal
+Introduce an automatic, opt-out mechanism for connected MicroShift clusters to
+report their status to Red Hat. Each MicroShift cluster should send the
+following metrics (at least) to the Telemetry API:
+* Number of available cores/RAM/disk
+* Average utilization of cores/RAM/disk (% of total)
+* Number of namespaces
+* Number of running pods
+* Number of container images on disk
+* Number of routes/ingress/services
+* Number of CRDs
+* OS version / Type (rpm / ostree)
+
+These metrics should be sent at least once a day.
+
+### Metrics details
+Based on [Telemetry API](https://rhobs-handbook.netlify.app/products/openshiftmonitoring/telemetry.md/)
+MicroShift will send metrics using Prometheus format. There are several ways
+of using the API: [using the client to forward metrics](https://github.com/openshift/telemeter/tree/main?tab=readme-ov-file#upload-endpoint-receive-metrics-in-client_modelmetricfamily-format-from-telemeter-client-currently-used-by-cmo),
+or [direct write requests](https://github.com/openshift/telemeter/tree/main?tab=readme-ov-file#metricsv1receive-endpoint-receive-metrics-in-prompbwriterequest-format-from-any-client).
+Using the client is not an option because that requires a local Prometheus
+instance, not viable in the typical resource constrained MicroShift deployment.
+The direct write requests take raw Prometheus data and require both special
+labels (`_id` for the cluster id) and crafting the authentication headers
+(cluster id and pull secret in a specific format in HTTP headers).
```

Comment:
it would be good to assess the potential impact on the Telemetry write endpoint (`https://infogw.api.openshift.com/metrics/v1/receive`).

Reply:
You mean in bandwidth used per day? Number of connections?

Reply:
@moadz would have more insights but my concerns would mainly be about network traffic and authentication requests.

Reply:
Included an [estimation](https://github.com/openshift/enhancements/pull/1742/files#diff-3993b301bc7e4a4e54d4f5f8f974ae071bd7406ae66bb981e72c79187dd2dbceR177-R181) of data sent per metrics payload/report.